### ISSUE TITLE
[feature] universal fern

### ISSUE THREAD

**Issue by ydzhou** [lambdalisue/fern.vim#402]
I would like to see an universal fern similar to what is in this [PR](https://github.com/lambdalisue/fern.vim/pull/261)

We will have essentially one fern across all the buffers for this vim instance.

**Comment by ydzhou** [comment]
Hi I thought that PR is enough to bring back the universal fern. What is missing here?

I am willing to help to fill the gap :)

**Comment by lambdalisue** [comment]
At that time, there is no right side drawer. That's why the PR probably does not support it.

**Comment by ydzhou** [comment]
Hi Alisue, can you take a look at the PR?

The main idea is to preserve a global copy of fern instance and always retrieve it for all the buffers. But I am not sure how vimscript handle the assignment of buffer local variable. If it is a deep copy, then maybe I need to find another way to avoid unnecessary memory wasting.

**Comment by ydzhou** [comment]
Looks like we are allowing both left and right drawer displayed in the same window. I am not sure the purpose of that.

I create a new PR to treat left and right drawer as separate thing. So we will have one universal drawer on the left and on the right.

Personally I think the `right` feature is confusing. My proposal is to keep only one fern drawer in one window. If you open it on the left side, then when you open it on the right side. The same left drawer will get redrawed to the right side. This can also make supporting `top` drawer or `bottom` drawer (if any user want to have) easier and clearer.

What do you think about it?

**Comment by lambdalisue** [comment]
> Looks like we are allowing both left and right drawer displayed in the same window. I am not understanding the purpose of that. But following this design choice, I create a new PR to treat the left and right drawer as two separate things. We will have one universal drawer on the left and another one on the right.

Because fern is not just a file explorer but a general purpose tree viewer. User may want to see JSON structure or [vim-iced-fern-debugger](https://github.com/liquidz/vim-iced-fern-debugger) or whatever on the right side, addition to the file explorer on the left side.

> Personally I think the -right feature is confusing. My proposal is to keep only one fern drawer in one window.
For example, if I run Fern . -drawer, it creates a drawer at the left side. Then I run Fern . -drawer -right, it close the existing left drawer and redraw it on the right side.

If the behavior is an opt-in feature, I think it's fair enough. However, that's a bit out of the topic so the PR should be separated if you'll do it.

**Comment by ydzhou** [comment]
> > Looks like we are allowing both left and right drawer displayed in the same window. I am not understanding the purpose of that. But following this design choice, I create a new PR to treat the left and right drawer as two separate things. We will have one universal drawer on the left and another one on the right.
> 
> Because fern is not just a file explorer but a general purpose tree viewer. User may want to see JSON structure or [vim-iced-fern-debugger](https://github.com/liquidz/vim-iced-fern-debugger) or whatever on the right side, addition to the file explorer on the left side.

Great to hear that Fern can do many more things!

Still I think, drawer concept is a little confusing to me. The reason why I discuss drawer here is because that I feel like current code has many hardcode parts to handle `left` and `right` drawer separately, yet I do not see much uniqueness of those two. types of drawers. I am wondering if we can make code more generic.

The difference between a drawer and a Fern tree is that a drawer is unique per tab, thus it can be toggled and its content can be perserved. Is that correct? Should we just introduce a type of Fern tree called `persistent tree` or whatever name you prefer and make it generic?

**Comment by lambdalisue** [comment]
I'm sorry but I couldn't find the benefit of your suggestion. "Making two options generic" is not enough for me to apply massive changes to the current code.

**Comment by ydzhou** [comment]
Yeah not much actual customer benefits, it is more of a refactoring so that we can support more cases easier. But I understand the return on investment for this work is very uncertain and we do not need to proceed unless you are seeing a clear benefit.

**Comment by ydzhou** [comment]
Hi Alisue

I am thinking to introduce a feature, it looks like this. When you open a drawer, it will stay opened in other tabs. When you close a draw, it will stay closed in other tabs.

For example,
I toggle the fern in this tab, then when I switch to other tabs, fern is already opened.
If I toggle the fern to close, then when I switch to other tabs, there will be no fern drawer.

This feature makes fern behave similar to sidebar for file tree in some IDEs.

How do you like it?

**Comment by lambdalisue** [comment]
It's interesting but a bit different from the feature described in this issue I think. Please create a new feature request for clarify the feature.
